src/toil/test/src/multipartTransferTest.py

```diff
@@ -0,0 +1,108 @@
+from toil.test import ToilTest, make_tests
+from toil.jobStores.aws.jobStore import copyKeyMultipart
+
+
+partSize = 2**20 * 5
+
+
+@contextmanager
+def openS3(keySize=None):
+    """
+    Creates an AWS bucket. If keySize is given a key of random bytes is created and its handle
+    is yielded. If no keySize is given an empty bucket handle is yielded. The bucket and all
+    created keys are cleaned up automatically.
+
+    :param int keySize: Size of source key.
+    """
+    if keySize < 0 and keySize is not None:
+        raise ValueError("Key size must be greater than zero")
+    with closing(boto.connect_s3(region='us-west-2')) as s3:
+        bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4(),
+                                  location=region_to_bucket_location('us-west-2'))
+        keyName = uuid.uuid4()
+        if keySize is None:
+            yield bucket
+        else:
+            def _uploadPart(partIndex):
+                if exceptions:
+                    return None
+                else:
+                    try:
+                        part = io.BytesIO(f.read(partSize))
+                        return upload.upload_part_from_file(fp=part, part_num=partIndex+1)
+                    except Exception as e:
+                        if partIndex <= 4:
+                            exceptions[partIndex] = traceback.format_exc()
+                        else:
+                            exceptions[partIndex] = e.message
+                        return None
+
+            totalSize = keySize
+            totalParts = (totalSize + partSize - 1) / partSize
+            exceptions = collections.OrderedDict()
+            upload = s3.get_bucket(bucket).initiate_multipart_upload(keyName)
+            try:
+                with open('/dev/urandom', 'rb') as f:
+                    with ThreadPoolExecutor(max_workers=min(cpu_count()*16, totalParts, 128)) as ex:
+                        finishedParts = set(ex.map(_uploadPart, range(0, totalParts)))
+                        if exceptions:
+                            raise RuntimeError()
+                        assert len(finishedParts) == totalParts
+                upload.complete_upload()
+            except:
+                with panic():
+                    upload.cancel_upload()
+                raise
+            yield bucket.get_key(keyName)
+
+        for key in bucket.list():
+            key.delete()
+        bucket.delete()
+
+
+@skip("too slow")  # Multipart copy is implicitly tested in import export tests.
+class AWSMultipartCopyTest(ToilTest):
+    def setUp(self):
+        super(AWSMultipartCopyTest, self).setUp()
+
+    @classmethod
+    def makeTests(self):
+        def _multipartCopy(self, threadPoolSize):
+            with openS3(threadPoolSize * partSize) as srcKey:
+                with openS3() as dstBucket:
+                    copyKeyMultipart(srcKey, dstBucket, "test")
+                    self.assertEqual(srcKey.get_contents_as_string(),
```

Comment:
This line here makes the complicated multi-threaded upload in openS3() look ridiculous.